### ISSUE TITLE
[Apple's Broken Compression Handling] NSPOSIXError when trying to connect from Damus

### ISSUE THREAD

**Issue by syobocat** [Yonle/blyat#3]
Hello.
When I try to connect to a bouncer from Damus app, it fails with `Error: Error Domain=NSPOSIXErrorDomain Code=100 "Protocol error" UserInfo={NSDescription=Protocol error}` and Damus attempts to reconnect over and over.

Here is the screenshot:
<img width="1136" alt="スクリーンショット 2024-07-28 20 27 52" src="https://github.com/user-attachments/assets/438cf144-7796-404b-bcbf-ff5e8431e388">

My bostr2 instance is running with Go 1.21.12 on 14.1-RELEASE-p2, and is behind Caddy and Cloudflare.

**Comment by Yonle** [comment]
This is strange. Try paste logs printed on the bostr2 server.

**Comment by Yonle** [comment]
Just checked, The error 100 is referring to failed HTTP/2 connections.

It looks like it's trying to connect with HTTP/2, which is not supported by [nhooyr.io/websocket](https://github.com/nhooyr/websocket#Roadmap) lib yet
![Screenshot_20240728-202708.jpg](https://github.com/user-attachments/assets/cf64432b-0a2e-4449-92e3-dbe722e62e2b)

**Comment by Yonle** [comment]
You could try restart bostr2 with `GODEBUG=http2server=0` environment variable being set.

**Comment by syobocat** [comment]
I set `GODEBUG=http2server=0` and restarted bostr2, but I still got the same error.

Also, I tried connecting to the server directly (without any reverse proxy), but the result was the same, too.

<img width="1136" alt="スクリーンショット 2024-07-28 23 29 05" src="https://github.com/user-attachments/assets/04bc1e72-2d96-4b52-8ae2-4e8bf9b46780">

Here is the log bostr2 printed:
```
⋮
2024/07/28 22:54:57 172.70.122.72 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 22:54:57 172.70.122.72: failed to get reader: failed to read frame header: EOF
2024/07/28 22:54:57 172.70.122.72 disconnect (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 22:54:59 162.158.118.73 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 22:55:00 162.158.118.73: failed to get reader: failed to read frame header: EOF
2024/07/28 22:55:00 162.158.118.73 disconnect (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 22:55:02 172.68.118.227 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 22:55:02 172.68.118.227: failed to get reader: failed to read frame header: EOF
2024/07/28 22:55:02 172.68.118.227 disconnect (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 22:55:04 172.70.222.139 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
⋮
```

If I try connecting directly, it shows a slightly different message:
```
⋮
2024/07/28 23:32:16 192.168.0.221:52741 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 23:32:16 192.168.0.221:52741: failed to get reader: received close frame: status = StatusProtocolError and reason = ""
2024/07/28 23:32:16 192.168.0.221:52741 disconnect (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 23:32:18 192.168.0.221:52743 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 23:32:18 192.168.0.221:52743: failed to get reader: received close frame: status = StatusProtocolError and reason = ""
2024/07/28 23:32:18 192.168.0.221:52743 disconnect (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 23:32:19 192.168.0.221:52744 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 23:32:19 192.168.0.221:52744: failed to get reader: received close frame: status = StatusProtocolError and reason = ""
2024/07/28 23:32:19 192.168.0.221:52744 disconnect (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
2024/07/28 23:32:21 192.168.0.221:52745 connected (damus/14 CFNetwork/1496.0.7 Darwin/23.5.0)
⋮
```

**Comment by Yonle** [comment]
Majority of servers out here are using HTTP/1.1, and they worked most of the time.

I checked the connection, Nothing wrong with header nor status.

Just added an option to stop serving invalid `Upgrade` header. I have no idea whenever it would made any difference.

Either way, You could test with the recent commit on master. Clone this repository, `go install .`, try restart and see will it make any difference.

If still persist, and this problem is not able to be reproduced on [bostr](https://github.com/Yonle/bostr), Then my final guessing is the problem is on nhooyr's websocket lib.

**Comment by Yonle** [comment]
Looks like it has something to do with compression.

https://github.com/nhooyr/websocket/issues/218

**Comment by Yonle** [comment]
Disabled compression due to it being unreliable at very certain cases. Could you try again with the recent commit?

**Comment by Yonle** [comment]
Added a code that disables compression on apple while keeping it on for others.

Well well well.

Right now, It should've fixed. But not a good fix unfortunately.

**Comment by Yonle** [comment]
Fixed on v1.0.1

**Comment by syobocat** [comment]
I tried 1.0.1 and it worked as expected.
Thank you so much!
